### PR TITLE
parallel migration

### PR DESCRIPTION
**Pull request by gites**
This PR speed up keys migration using the `vault operator migrate` command.
Each key is migrated in a goroutine (the actual part that moves the data) and thus the blocking operation of copying data is handled separately from the part that iterates over objects in Vault. In addition, the number of parallel goroutines migrating the data can be controlled by setting the `max_parallel` option in the migration config. 

Thanks to this change I manage to cut down Vault storage migration time between Consul and GCS from ~3h 30m to ~5m for around 90k of entries and `max_parallel=100`. 

If that looks promising I will update the docs too.

### CODE REVIEW

**Comment by hsimon-hashicorp** [comment]
> @hsimon-hashicorp what is the prospect of getting this one merged ?

I'll see what I can do! I've asked for a review from our engineer on rotation - please feel free to ping me again if need be, as well. :)

File: command/operator_migrate.go
Comment (HridoyRoy) [inline command/operator_migrate.go:48]: `max_parallel` might cause a bug with the `start` flag. Could we please make `max_parallel` a flag instead of a config option and error if both these flags are specified?

File: command/operator_migrate_test.go
Comment (HridoyRoy) [inline command/operator_migrate_test.go:59]: I'm not sure I follow how this tests the parallelism feature. Could you please elaborate on this?

File: command/operator_migrate.go
Comment (HridoyRoy) [inline command/operator_migrate.go:387]: I think this error doesn't make it to the outside `dfsScan ` function. Something should be done with the error.

**Review by HridoyRoy — CHANGES_REQUESTED** [review]
Hi, thanks for submitting this PR!

I've added some review comments. I think a one abstract/general question I have is around testing. There's a duplicate PR here https://github.com/hashicorp/vault/pull/8077/ for this feature, and it would be great if we could port some of the testing logic to the feature branch, or alternatively add some custom testing that would verify the parallelism itself is working as intended. 
Thanks again!

**Comment by hoerup** [comment]
@gites are you still active, and would you be willing on answering the questions?

File: command/operator_migrate.go
Comment (gites) [reply command/operator_migrate.go]: Removed in favor of a command flag.

File: command/operator_migrate_test.go
Comment (gites) [reply command/operator_migrate_test.go]: Hard to say after 2,5 years ;). Don't look like this makes any sens here. I set the parallel value to 10 only for DFS test and rest is set to 1.

File: command/operator_migrate.go
Comment (gites) [reply command/operator_migrate.go:414]: What would you suggest? Sending via channel?

**Comment by gites** [comment]
@HridoyRoy @hoerup I updated the pr and answered some questions.

File: command/operator_migrate.go
Comment (peteski22) [inline command/operator_migrate.go:147]: Please could we add some sanity check on the `c.flagParallel` value before this to ensure it's `>= 1`? You could modify this `if` statement to check `c.flagParallel > 1` after that.

File: command/operator_migrate.go
Comment (peteski22) [inline command/operator_migrate.go:102]: If we're doing checks at the start of the func, then it's not so much of an issue, but for clarity perhaps you could consider using `UintVar` as this would eliminate anyone doing something silly like passing `-parallel=-10` etc.

File: command/operator_migrate.go
Comment (peteski22) [inline command/operator_migrate.go:243]: `uint` could propagate through these methods.

File: command/operator_migrate.go
Comment (peteski22) [inline command/operator_migrate.go:226]: You'll probably need `int(c.flagParallel)` now if we move to `parallel` being an unsigned int.

File: command/operator_migrate.go
Comment (HridoyRoy) [reply command/operator_migrate.go:414]: That could work, or using an ErrorGroup should work I think. I can't think of any particular standardization at the moment that I'm particularly partial toward, just that there should be some error handling here. I can circle back with some other folks to see if there's a particular preference or standardization we have as well.

File: command/operator_migrate.go
Comment (mpalmi) [inline command/operator_migrate.go:147]: Can we use `math.MaxInt32` instead of a magic number (here and below)?

Comment (mpalmi) [reply command/operator_migrate.go:147]: As a matter of fact, we'll likely not gain much from running `math.maxInt32` goroutines in parallel and it may not even be possible. How do you feel about limiting `flagParallel` to some lower number. Say, 100?

File: command/operator_migrate.go
Comment (mpalmi) [inline command/operator_migrate.go:106]:
```suggestion
		Usage: "Use go routines when migrating. Can speed up the migration " +
```

File: command/operator_migrate.go
Comment (mpalmi) [reply command/operator_migrate.go:147]: This should actually be fine as just `math.maxInt32`. We can leave it up to the user to properly tune, but maybe we should specifically call that out in the [docs](https://github.com/hashicorp/vault/blob/main/website/content/docs/commands/operator/migrate.mdx#usage).

**Comment by msheldyakov** [comment]
Hello! I would like to finish this PR(it looks like almost everything is already here, there are little things left). @gites what do you think? and thanks for all this work!

**Comment by gites** [comment]
@msheldyakov go ahead and finish it! Good luck!

File: command/operator_migrate.go
Comment (peteski22) [reply command/operator_migrate.go:106]:
```suggestion
Usage: "Use goroutines when migrating. Can speed up the migration " +
```

**Comment by hghaf099** [comment]
@gites and @msheldyakov There are some comments and questions that we would like to sort out before merging the PR. Would you be interested to respond/address those comments?

**Comment by peteski22** [comment]
@gites I tried to help you out with some things if you've been too busy to get back to this, appreciate it's been a long lived affair.

https://github.com/gites/vault/pull/1

**Comment by hsimon-hashicorp** [comment]
Hi folks, are we still working on this? Otherwise, our engineers may be able to fork it and make some progress and get it merged. Thanks in advance!

**Comment by peteski22** [comment]
> Hi folks, are we still working on this? Otherwise, our engineers may be able to fork it and make some progress and get it merged. Thanks in advance!

Hey @hsimon-hashicorp, it may well be that the original contributor just hasn't the time now to continue this and it would be a shame to see their efforts wasted. I'm happy to push my fork of @gites' work into a HashiCorp branch so we can try and wrap it up. 

I'll leave this comment up for a few days in case @gites and other co-authors want to finish it themselves, and failing a response, or a response to suggest we 'go ahead' I think we should do just that. :)

**Comment by peteski22** [comment]
> > Hi folks, are we still working on this? Otherwise, our engineers may be able to fork it and make some progress and get it merged. Thanks in advance!
> 
> Hey @hsimon-hashicorp, it may well be that the original contributor just hasn't the time now to continue this and it would be a shame to see their efforts wasted. I'm happy to push my fork of @gites' work into a HashiCorp branch so we can try and wrap it up.
> 
> I'll leave this comment up for a few days in case @gites and other co-authors want to finish it themselves, and failing a response, or a response to suggest we 'go ahead' I think we should do just that. :)

Oh I see that someone else already asked and was granted permission to finish this off (in the comments above) https://github.com/hashicorp/vault/pull/8857#issuecomment-1313912890. Since nothing is happening on the PR I will start trying to move into 'into' HashiCorp.

**Comment by gites** [comment]
@hsimon-hashicorp @peteski22 Yes, go ahead

**Pull request closed without merging**
